src/lazy/expanded/compiler.rs

```diff
@@ -631,61 +626,84 @@ impl TemplateCompiler {
+        // surface an appropriate error message.
+        let macro_id = match id_expr {
+            None => {
+                return IonResult::decoding_error(
+                    "found an empty s-expression in an unquoted context",
+                )
+            }
+            Some(Err(e)) => return Err(e),
             Some(Ok(value)) => match value.read()? {
                 ValueRef::Symbol(s) => {
                     if let Some(name) = s.text() {
-                        let address =
-                            context.macro_table.address_for_name(name).ok_or_else(|| {
-                                IonError::decoding_error(format!("unrecognized macro name: {name}"))
-                            })?;
-                        Ok((Some(name.to_string()), address))
+                        MacroIdRef::LocalName(name)
                     } else {
-                        IonResult::decoding_error("macro names must be an identifier")
+                        return IonResult::decoding_error("macro names must be an identifier");
                     }
                 }
                 ValueRef::Int(int) => {
                     let address = usize::try_from(int.expect_i64()?).map_err(|_| {
                         IonError::decoding_error(format!("found an invalid macro address: {int}"))
                     })?;
-                    if context.macro_table.macro_at_address(address).is_none() {
-                        IonResult::decoding_error(format!(
-                            "invocation of invalid macro address {address}"
-                        ))
-                    } else {
-                        Ok((None, address))
-                    }
+                    MacroIdRef::LocalAddress(address)
+                }
+                other => {
+                    return IonResult::decoding_error(format!(
+                        "expected a macro name (symbol) or address (int), but found: {other:?}"
+                    ))
                 }
-                other => IonResult::decoding_error(format!(
-                    "expected a macro name (symbol) or address (int), but found: {other:?}"
-                )),
             },
+        };
+        // Now that we have our ID (name/address), try to resolve it to a previously defined macro.
+        // We'll look in the pending macros table first...
+        Self::resolve_macro_id_in(macro_id, tdl_context.pending_macros)
+            // ...or we'll raise an error.
+            .ok_or_else(|| {
+                IonError::decoding_error(format!("macro '{macro_id}' has not been defined (yet?)"))
+            })
+        // TODO: Should it be possible for a macro to reference (and therefore "close over") a macro
+        //       in the current context? The referenced macro would not necessarily be accessible
+        //       in the new encoding context, but would still need to be accessible internally.
```

Comment:
This would be consistent with how the symbol table works, right?

Reply:
Yes, in that the active symbol table remains available while reading a new symbol table. However, symbols don't depend on other symbol definitions (you can't put `$10` in the `symbols` list of an LST), so this is new territory in that respect.